## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [224]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [225]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [226]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [227]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [228]:
# What three item categories had the most entries?
top_categories = df["category"].value_counts()
top_categories.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [229]:
# For the category with the most entries,
# which subcategory had the most entries?
top_subcategory = df["subcategory"].value_counts()
top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [230]:
# Which five clients had the most entries in the data?
top_clients = df["client_id"].value_counts().head(5)
top_clients.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [231]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_ids = df["client_id"].value_counts().head(5).index.tolist()
top_five_clients_ids

[33615, 66037, 46820, 38378, 24741]

In [250]:
top_five_df = df[df['client_id'].isin(top_five_clients_ids)]
top_five_df

first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  client_id  order_id  \
5            5016915923         jessica.reyes@grimes.net      33615   1327730   
12           5016915923         jessica.reyes@grimes.net      33615   6059361   
137    444.233.9146x233            bmyers3696@wilson.com      46820   7109502   
214    444.233.9146x233            bmyers3696@wilson.com      46820   8967498   
216    729-620-6192x875           aeverett@rodriguez.com      66037   1203693   
...                 ...                              ...        ...       ...   
54535  729-620-6192x875           aeverett@rodriguez.com      66037   7550600   
54536      808.257.2029  garrettkendra8479@jefferson.com      24741   1886506   
54579  265-666-3155x118           ayoung9158@stevens.com      38378   8661147   
54589  265-666-3155x118           ayoung9158@stevens.com      38378   4508675   
54613      808.257.2029  garrettkendra8479@jefferson.com      24741   9114777   

       order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5      2023-05-02          18        2023  ...    154.44        2.79   78   
12     2023-05-16          20        2023  ...    198.38       27.51  568   
137    2023-05-12          19        2023  ...    111.47        5.65  118   
214    2023-02-06           6        2023  ...      6.26        7.05  133   
216    2023-04-09          14        2023  ...    166.09        0.26  157   
...           ...         ...         ...  ...       ...         ...  ...   
54535  2023-02-24           8        2023  ...     10.66        7.52  108   
54536  2023-05-26          21        2023  ...     40.05        4.70  721   
54579  2023-03-09          10        2023  ...     20.29       13.07  118   
54589  2023-03-25          12        2023  ...     38.26        0.70  685   
54613  2023-02-13           7        2023  ...     59.32       13.13   91   

       line_number  line_subtotal  total_weight  shipping_price  \
5                0       13481.52        217.62         1523.34   
12               2      130492.32      15625.68       109379.76   
137              9       16063.34        666.70         4666.90   
214              3        1061.34        937.65         6563.55   
216              2       47393.59         40.82          408.20   
...            ...            ...           ...             ...   
54535            2        1810.08        812.16         5685.12   
54536            3       46014.22       3388.70        23720.90   
54579            3        3687.50       1542.26        10795.82   
54589            0       35784.40        479.50         3356.50   
54613            2        8705.97       1194.83         8363.81   

          line_price  line_cost   line_profit  
5       16392.809550   13569.66   2823.149550  
12     262060.247400  222059.60  40000.647400  
137     22647.787200   17820.36   4827.427200  
214      8330.192325    7396.13    934.062325  
216     52223.455575   26484.33  25739.125575  
...              ...        ...           ...  
54535    8188.506000    6836.40   1352.106000  
54536   76185.618600   

In [274]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_qty = df.loc[(df["client_id"] == 33615, 'qty')].sum()
top_client_qty

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [233]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
subtotal_df = df[['unit_price', 'qty', 'line_subtotal']]
subtotal_df.head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [234]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
shipping_price_df = df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']]
shipping_price_df.head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [235]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['line_price'] = ((df['line_subtotal'] + df['shipping_price']) * .0925) + (df['line_subtotal'] + df['shipping_price'])
total_price_df = df[['line_subtotal', 'shipping_price', 'line_price']]
total_price_df.head()

line_subtotal  shipping_price     line_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600
3        1056.18          356.70    1543.571400
4        3902.00         6500.20   11364.403500

In [236]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   
3            3        1056.18         35.67          356.70    1543.571400   
4            1        3902.00        928.60         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [237]:
# Create a column for the profit of
# each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   
3        1056.18         35.67          356.70    1543.571400    1077.35   
4        3902.00        928.60         6500.20   11364.403500    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [238]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['line_price'].sum()
print("Calculated Total Prices:")
print(order_totals.loc[[2742071, 2173913, 6128929]])

Calculated Total Prices:
order_id
2742071    152811.897075
2173913    162388.719300
6128929    923441.240450
Name: line_price, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [242]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
client_spending = df.groupby('client_id')['line_price'].sum()
top_five_client_spending = client_spending.loc[top_five_clients_ids]
formatted_spending = top_five_client_spending.round(2)
print("Total spending by top 5 clients:")
print(formatted_spending)

Total spending by top 5 clients:
client_id
33615     8377308.57
66037    10259514.80
46820     9743794.32
38378    12906550.88
24741    82268891.98
Name: line_price, dtype: float64


In [296]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = top_five_df.groupby('client_id').sum('line_profit')
summary_df = summary_df.sort_values('line_profit', ascending=False)
summary_df = summary_df[['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']]
summary_df

qty  shipping_price    line_price    line_cost   line_profit
client_id                                                                 
24741      239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07
38378       73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
66037       43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
46820       75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
33615       64313      1828984.89  8.377309e+06   6175313.91  2.201995e+06

In [297]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
renamed_summary_df = summary_df.rename(columns={"qty": "Units", "shipping_price": "Shipping", "line_price": "Total Revenue", "line_cost": "Total Cost", "line_profit": "Total Profit"})
renamed_summary_df

Units    Shipping  Total Revenue   Total Cost  Total Profit
client_id                                                              
24741      239862  5126448.37   8.226889e+07  45688899.71  3.657999e+07
38378       73667  3429455.40   1.290655e+07   9634720.98  3.271830e+06
66037       43018  1395151.85   1.025951e+07   7004482.98  3.255032e+06
46820       75768  1601448.84   9.743794e+06   7007191.64  2.736603e+06
33615       64313  1828984.89   8.377309e+06   6175313.91  2.201995e+06

In [299]:
def format_dollars(number):
    return "${:,.2f}".format(number)

columns_format = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
for column in columns_format:
    renamed_summary_df[column] = renamed_summary_df[column].map(format_dollars)

renamed_summary_df

Units       Shipping   Total Revenue      Total Cost  \
client_id                                                          
24741      239862  $5,126,448.37  $82,268,891.98  $45,688,899.71   
38378       73667  $3,429,455.40  $12,906,550.88   $9,634,720.98   
66037       43018  $1,395,151.85  $10,259,514.80   $7,004,482.98   
46820       75768  $1,601,448.84   $9,743,794.32   $7,007,191.64   
33615       64313  $1,828,984.89   $8,377,308.57   $6,175,313.91   

             Total Profit  
client_id                  
24741      $36,579,992.27  
38378       $3,271,829.90  
66037       $3,255,031.82  
46820       $2,736,602.68  
33615       $2,201,994.66

In [300]:
# Sort the updated data by "Total Profit" form highest to lowest
renamed_summary_df

Units       Shipping   Total Revenue      Total Cost  \
client_id                                                          
24741      239862  $5,126,448.37  $82,268,891.98  $45,688,899.71   
38378       73667  $3,429,455.40  $12,906,550.88   $9,634,720.98   
66037       43018  $1,395,151.85  $10,259,514.80   $7,004,482.98   
46820       75768  $1,601,448.84   $9,743,794.32   $7,007,191.64   
33615       64313  $1,828,984.89   $8,377,308.57   $6,175,313.91   

             Total Profit  
client_id                  
24741      $36,579,992.27  
38378       $3,271,829.90  
66037       $3,255,031.82  
46820       $2,736,602.68  
33615       $2,201,994.66